plugins/aws_ecr/lib/samson_aws_ecr/samson_plugin.rb

```diff
@@ -3,29 +3,57 @@
 module SamsonAwsEcr
   class Engine < Rails::Engine
     AMAZON_REGISTRY = /\A.*\.dkr.ecr.(?<region>[\w\-]+).amazonaws.com\z/
-    mattr_accessor :ecr_client
-    mattr_accessor(:credentials_expire_at) { Time.current }
 
-    initializer :ecr_client do
-      if match = AMAZON_REGISTRY.match(ENV['DOCKER_REGISTRY'])
-        SamsonAwsEcr::Engine.ecr_client = Aws::ECR::Client.new(region: match['region'])
+    class << self
+      # we make sure the repo exists so pushes do not fail
+      # - ignores if the repo already exists
+      # - ignores if the repo cannot be created due to permission problems
+      def ensure_repository(repository)
+        if ecr_client
+          ecr_client.create_repository(
+            repository_name: repository.split('/', 2).last
+          )
+        end
+      rescue Aws::ECR::Errors::RepositoryAlreadyExistsException
+        Rails.logger.info("Repository already exists")
+      rescue Aws::ECR::Errors::AccessDenied
+        Rails.logger.info("Not allowed to create repositories")
+      end
+
+      # aws credentials are only valid for a limited time, so we need to refresh them before pushing
+      def refresh_credentials
+        if credentials_stale?
+          authorization_data = ecr_client.get_authorization_token.authorization_data.first
+
+          self.credentials_expire_at = authorization_data.expires_at || raise("NO EXPIRE FOUND")
+          user, pass = Base64.decode64(authorization_data.authorization_token).split(":", 2)
+          ENV['DOCKER_REGISTRY_USER'] = user
+          ENV['DOCKER_REGISTRY_PASS'] = pass
+        end
+      rescue Aws::ECR::Errors::InvalidSignatureException
+        raise Samson::Hooks::UserError, "Invalid AWS credentials"
       end
-    end
-  end
-end
 
-Samson::Hooks.callback :before_docker_build do
-  begin
-    if SamsonAwsEcr::Engine.ecr_client && SamsonAwsEcr::Engine.credentials_expire_at < Time.current
-      authorization_token = SamsonAwsEcr::Engine.ecr_client.get_authorization_token
-      authorization_data  = authorization_token.authorization_data.first
+      private
 
-      SamsonAwsEcr::Engine.credentials_expire_at = authorization_data.expires_at || raise("NO EXPIRE FOUND")
-      user, pass = Base64.decode64(authorization_data.authorization_token).split(":")
-      ENV['DOCKER_REGISTRY_USER'] = user
-      ENV['DOCKER_REGISTRY_PASS'] = pass
+      attr_accessor :credentials_expire_at
+
+      def ecr_client
+        return @ecr_client if defined?(@ecr_client)
+        @ecr_client = if match = AMAZON_REGISTRY.match(Rails.application.config.samson.docker.registry)
+          Aws::ECR::Client.new(region: match['region'])
+        end
+      end
+
+      def credentials_stale?
+        ecr_client && (!credentials_expire_at || credentials_expire_at < Time.current)
```

Comment:
Why not refresh the credentials before using them? This way, there is no magic number that might sometimes fail and is always refreshed when needed and not otherwise